autoprotocol/builders.py

```diff
@@ -1738,6 +1746,146 @@ def blowout(self, volume, initial_z, flowrate=None):
+        ----------
+        transports : dict, optional
+            Dictionary of the transport parameters
+        mode : str, optional
+            Mode of dispense type
+
+        Returns
+        -------
+        str
+            Mode of dispense type
+
+        Raises
+        ------
+        TypeError
+            mode is not in str
+        ValueError
+            mode does not contain valid mode value
+        TypeError
+            liquid is not of str
+        ValueError
+            liquid_class is not a valid value
+        ValueError
+            liquid_class did not resolve to a mode
+        ValueError
+            multiple liquid_class exists in one LiquidHandle
+        """
+        # dict for default dispense mode for each liquid_class
+        liquid_class_to_dispense_mode = {
+            "air": "air_displacement",
+            "default": "air_displacement",
+            "viscous": "positive_displacement"
+        }
+        if mode:
+            if not isinstance(mode, str):
+                raise TypeError(
+                    "mode: {} must be in str".format(mode)
+                )
+            if mode not in self.dispense_modes:
+                raise ValueError(
+                    "mode: {} must be one of the valid modes: {}"
+                    "".format(mode, self.dispense_modes)
+                )
+            return mode
+
+        # get liquid_classes from transport input
+        liquid_classes = set([transport["mode_params"]["liquid_class"]
+                              for transport in transports])
+        # remove automatically added 'air' (e.g. from blowout steps) and None
+        # classes.
+        other_classes = liquid_classes - {"air", None}
+        if not other_classes:
+            return liquid_class_to_dispense_mode["air"]
+
+        modes = set()
+        for other_class in other_classes:
+            # resolve mode for different other_class classes (except for None).
+            if not isinstance(other_class, str):
+                raise TypeError(
+                    "liquid_class: {} is not of type str".format(other_class)
+                )
+            if other_class not in self.liquid_classes:
+                raise ValueError(
+                    "liquid_class: {} must be one of the valid"
+                    "classes: {} ".format(other_class, self.liquid_classes)
+                )
+            if other_class not in liquid_class_to_dispense_mode.keys():
+                raise ValueError(
+                    "liquid_class: {} did not resolve accordingly. If there "
+                    "is a new liquid_class, make sure dictionary: {}"
+                    "is updated.".format(
+                        other_class, liquid_class_to_dispense_mode
+                    )
+                )
+            modes.add(liquid_class_to_dispense_mode[other_class])
+
+        # return error if there are incompatible liquid_class a set of
+        # transports.
+        if len(set(modes)) > 1:
```

Comment:
is there any chance we have no modes in which case the below `modes[0]` will blow up?

Reply:
There shouldn't be.

1851:
```
        if not other_classes:
            return liquid_class_to_dispense_mode["air"]
```

Otherwise it goes into the for loop, where unless an exception is thrown, a mode is added for each iteration.

Reply:
Since `modes` can only get value by looking up dictionary, and there is an error thrown to catch such error, so there shouldn't be a case where `modes` will be empty at this point. I guess there used to be `if not modes: ...` for a different purpose before... I can add a check just in case.